### PR TITLE
General Windows improvements / c_* and winapi types

### LINKED ISSUE #39: Simulating scroll doesn't work on Windows
Changing the event in `test_listen_and_simulate()` to `EventType::Wheel{delta_x: 0, delta_y: -1}` (or delta_y: 1) gives a test failure because simulating wheel events does nothing! Print-debugging indicates that listening for scroll events with a hardware mouse works as expected. And finally, I wrote a snippet with the buggy behavior.
```rust
fn main() {
    loop {
        simulate(&EventType::Wheel {
            delta_x: 0,
            delta_y: -1,
        }).unwrap();
        std::thread::sleep(std::time::Duration::from_millis(100))
    }
}
```

details: Windows 10 version 1903, rdev master branch

### PR DESCRIPTION
**Pull request by maxbla**
fixes #32 for Windows only, fixes #39 #40

### CODE REVIEW

File: src/windows/simulate.rs [hunk 1]
```diff
@@ -58,69 +76,54 @@ fn keyboard_event(flags: u32, vk: u16, scan: u16) -> Result<(), SimulateError> {
...
             let width = unsafe { GetSystemMetrics(SM_CXVIRTUALSCREEN) };
             let height = unsafe { GetSystemMetrics(SM_CYVIRTUALSCREEN) };
 
-            mouse_event(
-                MOUSEEVENTF_MOVE | MOUSEEVENTF_ABSOLUTE,
+            sim_mouse_event(
+                MOUSEEVENTF_MOVE | MOUSEEVENTF_ABSOLUTE | MOUSEEVENTF_VIRTUALDESK,
                 0,
-                (*x * 65335.0) as i32 / width,
-                (*y * 65335.0) as i32 / height,
```
Comment (maxbla) [inline src/windows/simulate.rs]: typo! 65335 looks a lot like 65535. This took me longer to find than I'd like to admit.

File: src/windows/simulate.rs [hunk 2]
```diff
@@ -1,53 +1,71 @@
...
         Ok(())
     }
 }
 
-fn keyboard_event(flags: u32, vk: u16, scan: u16) -> Result<(), SimulateError> {
-    let mut input = INPUT {
+fn sim_keyboard_event(flags: DWORD, vk: WORD, scan: WORD) -> Result<(), SimulateError> {
+    let mut union: INPUT_u = unsafe { std::mem::zeroed() };
+    let inner_union = unsafe { union.ki_mut() };
+    *inner_union = KEYBDINPUT {
```
Comment (maxbla) [inline src/windows/simulate.rs]: while more verbose, this is how the winapi crate recommends you instantiate unions.

File: src/windows/keycodes.rs [hunk 3]
```diff
@@ -1,328 +1,138 @@
...
-const KP2: u16 = 98;
-const KP3: u16 = 99;
-const KP4: u16 = 100;
-const KP5: u16 = 101;
-const KP6: u16 = 102;
-const KP7: u16 = 103;
-const KP8: u16 = 104;
-const KP9: u16 = 105;
-const KP_DELETE: u16 = 110;
+macro_rules! decl_keycodes {
```
Comment (maxbla) [inline src/windows/keycodes.rs]: @Narsil how do you feel about the decl_keycode macro?

**Comment by maxbla** [comment]
This is failing CI because I added extra tests which don't pass on Ubuntu - I didn't actually change any linux-specific code, only expose an existing bug. I will still fix CI, but give me some time.

File: src/windows/keycodes.rs (hunk 3 above)
Comment (Narsil) [reply src/windows/keycodes.rs:5]: Seems much better. I'm usually against magic, but it's always good when it prevents us from shooting in the foot.

**Comment by maxbla** [comment]
Okay, now I have no clue why Ubuntu tests are failing. They pass on my local computer...

File: tests/listen_and_simulate.rs [hunk 4]
```diff
@@ -21,31 +24,50 @@ fn send_event(event: Event) {
...
 
-#[test]
-fn test_listen_and_simulate() {
+fn sim_then_listen(events: &mut dyn Iterator<Item = EventType>) -> Result<(), Box<dyn Error>> {
     // spawn new thread because listen blocks
     let _listener = thread::spawn(move || {
         listen(send_event).expect("Could not listen");
     });
+    let tenth_sec = Duration::from_millis(100);
+    thread::sleep(tenth_sec);
```
Comment (Narsil) [inline tests/listen_and_simulate.rs:33]: Try waiting for longer ? Maybe the github docker images are slower than this to start the listening ? I have 1 s set on my grab branch to get it working.

Comment (maxbla) [reply tests/listen_and_simulate.rs]: wow, that seems to have fixed it. Crazy. My rule of thumb is that 100ms is like 10 eternities for a computer. CI must be doing something weird that I'm not thinking of. Maybe Xvfb takes some time to initialize? Or maybe the CI server is virtualized on a host that is very busy. Regardless, I think this is ready to merge.

Comment (Narsil) [reply tests/listen_and_simulate.rs]: Perfect !

**Pull request merged**